README.md
# Harbormaster
Harbormaster is a [webhook handler](https://kubernetes.io/docs/admin/authentication/#webhook-token-authentication)
for validating JWT tokens sent to the Kubernetes API. This gives the flexibility
to add extra information to the user object for Kubernetes.

[![GoDoc](https://godoc.org/github.com/Nike-Inc/harbormaster?status.svg)](https://godoc.org/github.com/Nike-Inc/harbormaster)
[![Go Report Card](https://goreportcard.com/badge/github.com/Nike-Inc/harbormaster)](https://goreportcard.com/report/github.com/Nike-Inc/harbormaster)

## Usage
Harbormaster can be configured using command line flags or environment variables.
A list of those is below:

```
Flags:
  -r, --ca-cert string               ca certificate to use for root trust. This is required. [HARBORMASTER_CA_CERT] (default "/usr/share/harbormaster/ca.pem")
  -e, --cache-expiry duration        how long before a cache entry is considered invalid [HARBORMASTER_CACHE_EXPIRY] (default 1h0m0s)
  -m, --cache-max-memory int         amount of memory (in MB) to cap the cache at [HARBORMASTER_CACHE_MAX_MEMORY] (default 1024)
  -c, --cert string                  certificate to use for https. This is required. [HARBORMASTER_CERT] (default "/usr/share/harbormaster/harbormaster.pem")
  -i, --client-id string             OIDC provider client ID. [HARBORMASTER_CLIENT_ID]
      --graph-client-id string       MS Graph application client ID to use [HARBORMASTER_GRAPH_CLIENT_ID]
      --graph-client-secret string   MS Graph application client secret to use [HARBORMASTER_GRAPH_CLIENT_SECRET]
      --graph-tenant-name string     MS Graph application tenant name to use (generally <company>.onmicrosoft.com) [HARBORMASTER_GRAPH_TENANT_NAME]
  -g, --group-claim string           claim to use for user's groups [HARBORMASTER_GROUP_CLAIM] (default "groups")
      --health-port int              port to listen on for health checks. [HARBORMASTER_HEALTH_PORT] (default 8000)
  -h, --help                         help for harbormaster
  -u, --issuer-url string            OIDC token issuer URL. [HARBORMASTER_ISSUER_URL] (default "https://example.oktapreview.com")
  -k, --key string                   keyfile for the certificate. This is required. [HARBORMASTER_KEY] (default "/usr/share/harbormaster/harbormaster.key")
  -v, --log-level string             log level to use. Must be one of: 'DEBUG', 'INFO', 'WARN', or 'ERROR'. [HARBORMASTER_LOG_LEVEL] (default "WARN")
  -p, --port int                     port harbormaster should listen on. [HARBORMASTER_PORT] (default 9000)
  -n, --username-claim string        claim to use as the username [HARBORMASTER_USERNAME_CLAIM] (default "email")
```

Harbormaster relies on TLS authentication to validate the connection. We recommend
using the same CA that is used for the Kubernetes API server, and, if possible,
the same API server key for setting up HTTPS. HTTPS **is required** for the webhook
and we have no plans to support authentication webhooks without encryption.

### Group getters
Harbormaster supports getting additional user groups from other sources. Currently,
only one source is implemented. However, additional sources can be added by
implementing the `groups.Getter` interface. The groups sent back to Kubernetes
will be any groups contained in the JWT combined with any groups from a group getter.
If no group getter configuration is specified, Harbormaster will only use the 
groups from the JWT.

For performance, Harbormaster caches group lookups. The amount of time a cache 
entry is valid can be adjusted with a command line flag.

#### Microsoft Graph
You can optionally specify a Microsoft Graph API [client ID, secret](https://developer.microsoft.com/en-us/graph/docs/concepts/auth_v2_service),
and tenant name. If you specify these flags or environment variables, Harbormaster
will fetch groups from MS Graph.

### Caveats
The username claim specifies which username is used when querying the MS graph API. 
So if it isn't an email address, the lookup in the MS Graph API may fail.

### Docker images
We build Docker images for Harbormaster that are hosted on Docker Hub.
Every commit to master pushes a new build tagged as `canary` for those who want
to use the bleeding edge. Otherwise, the version of Harbormaster will be used to tag the image.

## Development
Development should be fairly straightforward for anyone who has built Go-based project. There are 
only 2 dependencies: a valid Go installation and the new [`dep` tool](https://github.com/golang/dep#setup) 
(that will soon be added into the main Go toolchain). Once you have those installed, 
you will need to clone this project into the proper place in your `GOPATH`:

```bash
mkdir -p $GOPATH/src/github.com/Nike-Inc
cd $GOPATH/src/github.com/Nike-Inc
git clone git@github.com:Nike-Inc/harbormaster.git
cd harbormaster
```

To get started, run `make bootstrap` to install project dependencies. Once that
is done, you can run `make test` to run the package tests and `make build` to
build a local binary at `./harbormaster`.

### How it works
Most of the functionality is broken out into different internal packages that
can only be used by Harbormaster. There are also a few external packages. The
internal packages are located at `./internal/pkg` and a description of each follows.

|   Package   |                                               Purpose                                               |
| ----------- | --------------------------------------------------------------------------------------------------- |
| `handler`   | Contains http handler functions and related helpers to handle and decode the `TokenReview` requests |
| `validator` | Contains all necessary functions for validating and caching token data                              |

The external packages currently consist only of code related to group getters:

| Package  |                                             Purpose                                             |
| -------- | ----------------------------------------------------------------------------------------------- |
| `graph`  | A package for logging into and accessing user group information from the MS Graph API           |
| `groups` | A simple package containing the `Getter` interface that must be implemented by any group getter |

The workflow can be a bit confusing at first glance. It is broken down below
for easy reference:

1. Harbormaster receives a JSON encoded `TokenReview` object from the API server
1. The `TokenReview` is deserialized into a native Kubernetes `TokenReview` object
1. The token is retrieved from the object and validated using an OIDC library
1. A map of claims is retrieved from the token for use in constructing the response to the API server
1. If a group getter is configured, these additional steps occur
    1. The validator checks if a list of user groups associated with the username exists
    1. If the cache entry exists, it uses those groups. Otherwise it uses the getter to fetch a group list and caches it
1. Using the username and group list, a new `TokenReview` object is created and serialized to JSON before sending back to the API server

### Releases and Docker images
To build binaries for a release, run the `make release` command. This will 
build binaries for Windows, Linux, and MacOS in parallel and tag them with a
version using `git describe --always --tags`. You can also manually specify a
version using the `VERSION` environment variable.

If you wish to build a docker image, you can use `make image` to build one. Essentially,
all that is happening is a `make release` and then it wraps those binaries in a
Docker image. Once again, `VERSION` will be used for tagging the image.

To push an image, run `make image-push`. You must be logged in to Docker Hub in
order to run this step. If you used `VERSION` in any of the previous steps, you
will need to set it here as well. 

## Roadmap
We keep track of outstanding issues and enhancement in the [issues page](https://github.com/Nike-Inc/harbormaster/issues).
You can follow along with existing issues or request new features there. However,
the first priorities on our list are below:

- Integration Tests
- Performance optimizations
- Distributed cache instead of a local cache

## Maintainers

- [Taylor Thomas](https://github.com/thomastaylor312)
- [Roger Ignazio](https://github.com/rji)
- [Jay Wallace](https://github.com/mootpt)


CHANGELOG.md
# Harbormaster Change Log

## v1.0.0
* Initial open source release


harbormaster_test.go
/*
Copyright 2017 Nike Inc.

Licensed under the Apache License, Version 2.0 (the License);
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an AS IS BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"io/ioutil"
	"os"
	"testing"
)

var tempCertRaw = `-----BEGIN CERTIFICATE-----
MIIDhDCCAmygAwIBAgIQLsBdzOEkh6k1P7uf5tDoDjANBgkqhkiG9w0BAQsFADBc
MQkwBwYDVQQGEwAxCTAHBgNVBAgTADEJMAcGA1UEBxMAMQkwBwYDVQQREwAxETAP
BgNVBAoTCGJvb3RrdWJlMQkwBwYDVQQLEwAxEDAOBgNVBAMTB2t1YmUtY2EwHhcN
MTcwNzI1MTkyNTE1WhcNMTgwNzI1MTkyNTE1WjBcMQkwBwYDVQQGEwAxCTAHBgNV
BAgTADEJMAcGA1UEBxMAMQkwBwYDVQQREwAxETAPBgNVBAoTCGJvb3RrdWJlMQkw
BwYDVQQLEwAxEDAOBgNVBAMTB2t1YmUtY2EwggEiMA0GCSqGSIb3DQEBAQUAA4IB
DwAwggEKAoIBAQCzcaXGF0MWjhW0E3LCW8d8dEvVBA5ownzaBvtq/1nxlNOVofEz
JOA8KePIVK2Nf4TcdE1X27lb+DRviX42Lmz8OR0zCDfhUEtClQIFU0E20xhYzpwE
SH3iiCjQotJRKyF889xcYCnVDwbeZAg5vSone6edCaFHfGrtT+ybhD9GMn13n0T9
cR8pvMgVvEPtiHWtvdTUXL3CTgkW8oYKQIk9C2ABf4cvEZsdNoPmDa/m4STRI776
DxxVzXINY9u+X/+GkLfZeemz2+jh/AOIQn8KLwQFd2R4iBXne3xKZhEQ5SYgiMFH
BsyCVnd1/0+FSzYSmngXXEt4YMy3jp2w2R4BAgMBAAGjQjBAMA4GA1UdDwEB/wQE
AwICpDAPBgNVHRMBAf8EBTADAQH/MB0GA1UdDgQWBBSmMMfemJRpFFNYW/gdua0e
St57+jANBgkqhkiG9w0BAQsFAAOCAQEAjk6cKuiYLqskDCfdJmhCz7U0T91CeL8N
9LDfXdcewcsSMg3PsJFEbNWriU9xtkQ5oaQy12XEyWxlV53UzSsAeFsQIAPi81zs
kHure/Ycg9lqV3GUz4ahzarrhd+irAhkHtTQCcy/irfCMR+jltg2hLazr97Y1/yF
4IPkQnP6jTR/s0c5gvmkuMk2ctsW0oe+xls5oglllXLjPvU0++kzZ+8CqK8CRcJw
jsIZMQi+TNG5HdZcDjV2nzh+wCo52kf4ShArgJRdDUeR18CPGJWhO0Efv236Iyv0
2mdECH7jZhL7/5LCSm+xteXoy1soxK3p4AI2VZsL7oRrgoGkj+etvA==
-----END CERTIFICATE-----
`

func createTmpCert(t *testing.T, content string) (string, func()) {
	f, err := ioutil.TempFile(os.TempDir(), "harbormaster-test")
	if err != nil {
		t.Fatal(err)
	}
	_, err = f.WriteString(content)
	if err != nil {
		t.Fatal(err)
	}
	f.Close()
	return f.Name(), func() {
		os.Remove(f.Name())
	}
}

func TestGenerateTLSConfig(t *testing.T) {
	t.Run("valid cert returns valid config", func(t *testing.T) {
		certPath, cleanup := createTmpCert(t, tempCertRaw)
		defer cleanup()
		conf, err := generateTLSConfig(certPath)
		if err != nil {
			t.Errorf("Did not expect error generating TLS config: %s", err)
		}
		if conf == nil {
			t.Error("Nil config was returned")
		}
	})

	t.Run("invalid cert returns error", func(t *testing.T) {
		certPath, cleanup := createTmpCert(t, "blah blah")
		defer cleanup()
		conf, err := generateTLSConfig(certPath)
		if err == nil {
			t.Error("Expected an error with an invalid certificate")
		}
		if conf != nil {
			t.Error("Returned config should have been nil and wasn't")
		}
	})

	t.Run("invalid file path returns error", func(t *testing.T) {
		conf, err := generateTLSConfig("a/totally/not/valid/path")
		if err == nil {
			t.Error("Expected an error with an invalid path")
		}
		if conf != nil {
			t.Error("Returned config should have been nil and wasn't")
		}
	})
}


pkg/graph/graph_test.go
/*
Copyright 2017 Nike Inc.

Licensed under the Apache License, Version 2.0 (the License);
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an AS IS BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package graph

import (
	"fmt"
	"net/http"
	"net/http/httptest"
	"net/url"
	"testing"
	"time"
)

func getAuthServerAndUserInfo(returnCode int, body, clientID, clientSecret string) (*httptest.Server, *UserInfo) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(returnCode)
		w.Write([]byte(body))
	}))
	loginURL, _ := url.Parse(ts.URL)
	u := &UserInfo{
		client:       http.DefaultClient,
		loginURL:     loginURL,
		headers:      http.Header{},
		clientID:     clientID,
		clientSecret: clientSecret,
	}
	return ts, u
}

func TestLogin(t *testing.T) {
	t.Run("successful login", func(t *testing.T) {
		var validToken = "blackbriar"
		var validBody = `{
  "token_type": "Bearer",
  "expires_in": 3599,
  "access_token": "%s"
}`
		ts, u := getAuthServerAndUserInfo(http.StatusOK, fmt.Sprintf(validBody, validToken), "jason", "bourne")
		defer ts.Close()

		err := u.login()
		if err != nil {
			t.Errorf("Error when trying to log in: %s", err)
		}
		if u.headers.Get("Authorization") != fmt.Sprintf("Bearer %s", validToken) {
			t.Errorf("Authorization header should be set. Expected: %q. Got: %q", fmt.Sprintf("Bearer %s", validToken), u.headers.Get("Authorization"))
		}
		if !time.Now().Before(u.expires) {
			t.Errorf("Expiry not set properly. Expected it to be after the current time. Actual: %v", u.expires)
		}
	})

	t.Run("unsuccessful login", func(t *testing.T) {
		ts, u := getAuthServerAndUserInfo(http.StatusUnauthorized, "Unauthorized", "CIA", "treadstone")
		defer ts.Close()

		err := u.login()
		if err == nil {
			t.Error("Should have gotten error")
		}
	})

	t.Run("request error", func(t *testing.T) {
		badURL, _ := url.Parse("https://127.0.0.1:34567")
		u := &UserInfo{
			client:       http.DefaultClient,
			loginURL:     badURL,
			headers:      http.Header{},
			clientID:     "CIA",
			clientSecret: "outcome",
		}

		err := u.login()
		if err == nil {
			t.Error("Should have gotten error")
		}
	})

	t.Run("bad response body", func(t *testing.T) {
		ts, u := getAuthServerAndUserInfo(http.StatusOK, "{bad_json", "CIA", "treadstone")
		defer ts.Close()

		err := u.login()
		if err == nil {
			t.Error("Should have gotten error")
		}
	})
}

func getAPIServerAndUserInfo(returnCode int, body string) (*httptest.Server, *UserInfo) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(returnCode)
		w.Write([]byte(body))
	}))
	apiURL, _ := url.Parse(ts.URL)
	u := &UserInfo{
		client:       http.DefaultClient,
		apiURL:       apiURL,
		headers:      http.Header{},
		clientID:     "jason",
		clientSecret: "bourne",
		expires:      time.Now().Add(time.Hour),
	}
	return ts, u
}

func TestGetGroupIDs(t *testing.T) {
	t.Run("successful request", func(t *testing.T) {
		var validBody = `{
  "value": [
      "f36ec2c5-fa5t-4f05-b87f-deadbeef"
  ]
}`
		ts, u := getAPIServerAndUserInfo(http.StatusOK, validBody)
		defer ts.Close()

		groups, err := u.getGroupIDs("john.michael.kane@yacht.io")
		if err != nil {
			t.Errorf("Should not have gotten error: %s", err)
		}
		if len(groups) != 1 {
			t.Errorf("Should have gotten a list of group IDs with 1 entry. Got: %d", len(groups))
		}
	})
	t.Run("bad server response", func(t *testing.T) {
		ts, u := getAPIServerAndUserInfo(http.StatusInternalServerError, "shutdown")
		defer ts.Close()

		groups, err := u.getGroupIDs("alexander.conklin@cia.gov")
		if err == nil {
			t.Error("Should have gotten error")
		}
		if groups != nil {
			t.Error("Group list should be nil")
		}
	})
	t.Run("request error", func(t *testing.T) {
		badURL, _ := url.Parse("https://127.0.0.1:34567")
		u := &UserInfo{
			client:       http.DefaultClient,
			apiURL:       badURL,
			headers:      http.Header{},
			clientID:     "jason",
			clientSecret: "bourne",
			expires:      time.Now().Add(time.Hour),
		}

		groups, err := u.getGroupIDs("richard.webb@cia.gov")
		if err == nil {
			t.Error("Should have gotten error")
		}
		if groups != nil {
			t.Error("Group list should be nil")
		}
	})
	t.Run("bad response body", func(t *testing.T) {
		ts, u := getAPIServerAndUserInfo(http.StatusOK, "{bad_json")
		defer ts.Close()

		groups, err := u.getGroupIDs("nicky.parsons@cia.gov")
		if err == nil {
			t.Error("Should have gotten error")
		}
		if groups != nil {
			t.Error("Group list should be nil")
		}
	})
}

func TestGetExpandedGroups(t *testing.T) {
	t.Run("successful request", func(t *testing.T) {
		var validBody = `{
  "value": [
    {
      "@odata.type": "#microsoft.graph.group",
      "displayName": "Treadstone.Assets.All"
    }
  ]
}`
		ts, u := getAPIServerAndUserInfo(http.StatusOK, validBody)
		defer ts.Close()

		groups, err := u.getExpandedGroups([]string{"f36ec2c5-fa5t-4f05-b87f-deadbeef"})
		if err != nil {
			t.Errorf("Should not have gotten error: %s", err)
		}
		if len(groups.Value) != 1 {
			t.Errorf("Should have gotten a list of groups with 1 entry. Got: %d", len(groups.Value))
		}
	})
	t.Run("bad server response", func(t *testing.T) {
		ts, u := getAPIServerAndUserInfo(http.StatusInternalServerError, "shutdown")
		defer ts.Close()

		groups, err := u.getExpandedGroups([]string{"f36ec2c5-fa5t-4f05-b87f-deadbeef"})
		if err == nil {
			t.Error("Should have gotten error")
		}
		if groups != nil {
			t.Error("Group list should be nil")
		}
	})
	t.Run("request error", func(t *testing.T) {
		badURL, _ := url.Parse("https://127.0.0.1:34567")
		u := &UserInfo{
			client:       http.DefaultClient,
			apiURL:       badURL,
			headers:      http.Header{},
			clientID:     "jason",
			clientSecret: "bourne",
			expires:      time.Now().Add(time.Hour),
		}

		groups, err := u.getExpandedGroups([]string{"f36ec2c5-fa5t-4f05-b87f-deadbeef"})
		if err == nil {
			t.Error("Should have gotten error")
		}
		if groups != nil {
			t.Error("Group list should be nil")
		}
	})
	t.Run("bad response body", func(t *testing.T) {
		ts, u := getAPIServerAndUserInfo(http.StatusOK, "{bad_json")
		defer ts.Close()

		groups, err := u.getExpandedGroups([]string{"f36ec2c5-fa5t-4f05-b87f-deadbeef"})
		if err == nil {
			t.Error("Should have gotten error")
		}
		if groups != nil {
			t.Error("Group list should be nil")
		}
	})
}

// This is only testing the full function run, error cases are handled in the tests above
func TestGetGroups(t *testing.T) {
	var validBody1 = `
{
    "value": [
        "f36ec2c5-fa5t-4f05-b87f-deadbeef"
    ]
}`
	var validBody2 = `{
	"value": [
		{
		    "@odata.type": "#microsoft.graph.group",
		    "displayName": "Treadstone.Assets.All"
		}
	]
}`
	mux := http.NewServeMux()
	mux.Handle("/users/", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(200)
		w.Write([]byte(validBody1))
	}))
	mux.Handle("/directoryObjects/", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(200)
		w.Write([]byte(validBody2))
	}))
	ts := httptest.NewServer(mux)
	apiURL, _ := url.Parse(ts.URL)
	u := &UserInfo{
		client:       http.DefaultClient,
		apiURL:       apiURL,
		headers:      http.Header{},
		clientID:     "jason",
		clientSecret: "bourne",
		expires:      time.Now().Add(time.Hour),
	}
	defer ts.Close()

	groups, err := u.GetGroups("blackbriar@cia.gov")
	if err != nil {
		t.Errorf("Should not have gotten error: %s", err)
	}
	if len(groups) != 1 {
		t.Errorf("Should have gotten a list of groups with 1 entry. Got: %d", len(groups))
	}
}


internal/pkg/handler/handler_test.go
/*
Copyright 2017 Nike Inc.

Licensed under the Apache License, Version 2.0 (the License);
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an AS IS BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package handler

import (
	"bytes"
	"fmt"
	"io"
	"io/ioutil"
	"net"
	"net/http"
	"net/http/httptest"
	"net/url"
	"strings"
	"testing"
	"time"

	"github.com/Nike-Inc/harbormaster/internal/pkg/validator"
	authenticationapi "k8s.io/api/authentication/v1beta1"
)

var hasNetwork bool

var oktaURL = url.URL{
	// We don't want to specify a port on here or else the issuer URL doesn't match
	// what the provider sends back (e.g. https://nike.okta.com vs https://nike.okta.com:443)
	Host:   "nike.okta.com",
	Scheme: "https",
}

// checkNetworkConnectivity checks to see if oktaURL is reachable before running a test
func checkNetworkConnectivity() {
	conn, err := net.DialTimeout("tcp", fmt.Sprintf("%s:443", oktaURL.Host), time.Second*3)
	if err != nil {
		// Explicitly setting here for clarity sake
		hasNetwork = false
		return
	}
	hasNetwork = true
	conn.Close()
}

func init() {
	checkNetworkConnectivity()
}

func getReaderFromString(data string) io.ReadCloser {
	return ioutil.NopCloser(bytes.NewReader([]byte(data)))
}

func TestNewTokenHandler(t *testing.T) {
	if !hasNetwork {
		t.Skipf("Unable to contact %s, skipping test", oktaURL.String())
	}
	v, err := validator.New(validator.IssuerURL(oktaURL.String()))
	if err != nil {
		t.Fatalf("Error trying to set up validator, exiting test: %s", err)
	}
	th := NewTokenHandler(v)
	if th == nil {
		t.Error("Got nil token handler")
	}
}

func TestParseTokenReview(t *testing.T) {
	var tokenReviewRequest = `{
  "apiVersion": "authentication.k8s.io/v1beta1",
  "kind": "TokenReview",
  "spec": {
    "token": "engage"
  }
}`

	var badTokenReviewRequest = `{ "bad_json":`
	t.Run("valid body should parse", func(t *testing.T) {
		body := getReaderFromString(tokenReviewRequest)
		review, err := parseTokenReview(body)
		if err != nil {
			t.Errorf("Got error while parsing TokenReview: %s", err)
		}
		if review == nil {
			t.Error("Got nil token review")
		}
	})
	t.Run("invalid object should error", func(t *testing.T) {
		body := getReaderFromString(badTokenReviewRequest)
		review, err := parseTokenReview(body)
		if err == nil {
			t.Error("Should have errored")
		}
		if review != nil {
			t.Error("Token review is not nil")
		}
	})
}

func TestHealthCheck(t *testing.T) {
	r := httptest.NewRequest(http.MethodGet, "http://127.0.0.1:8000/healthz", nil)
	w := httptest.NewRecorder()
	HealthCheck(w, r)
	resp := w.Result()
	if resp.StatusCode != http.StatusOK {
		t.Errorf("Expected a %d status, got %d instead", http.StatusOK, resp.StatusCode)
	}
}
func TestReviewError(t *testing.T) {
	if !hasNetwork {
		t.Skipf("Unable to contact %s, skipping test", oktaURL)
	}
	th := &TokenHandler{}
	var responseError = &authenticationapi.TokenReview{}
	var errorMessage = "warp drive failure"
	w := httptest.NewRecorder()
	th.reviewError(w, responseError, errorMessage, http.StatusInternalServerError)
	resp := w.Result()
	if resp.StatusCode != http.StatusInternalServerError {
		t.Errorf("Expected a %d status, got %d instead", http.StatusInternalServerError, resp.StatusCode)
	}
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		t.Errorf("Error while decoding body: %s", err)
	}
	strBody := string(body)
	if !strings.Contains(strBody, errorMessage) {
		t.Errorf("Response body does contain error. \nExpected to find: %s\nGot: %s", errorMessage, strBody)
	}
}


pkg/graph/types.go
/*
Copyright 2017 Nike Inc.

Licensed under the Apache License, Version 2.0 (the License);
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an AS IS BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package graph

// AuthResponse represents a response from the MS Graph auth API
type AuthResponse struct {
	TokenType string `json:"token_type"`
	Expires   int    `json:"expires_in"`
	Token     string `json:"access_token"`
}

// NOTE: These below are partial implementations of the API objects containing
// only the necessary fields to perform the functions of this package

// ObjectList represents a list of directory object IDs returned from the MS Graph API
type ObjectList struct {
	Value []string `json:"value"`
}

// ObjectQuery represents a query object to the directoryObjects endpoint
type ObjectQuery struct {
	IDs   []string `json:"ids"`
	Types []string `json:"types"`
}

// GroupList represents a list of groups returned from the MS Graph API
type GroupList struct {
	Value []Group `json:"value"`
}

// Group represents the Group object from the MSGraphAPI
type Group struct {
	Name string `json:"displayName"`
	ID   string `json:"id"`
}


harbormaster.go
/*
Copyright 2017 Nike Inc.

Licensed under the Apache License, Version 2.0 (the License);
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an AS IS BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"crypto/tls"
	"crypto/x509"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"strings"

	"github.com/Nike-Inc/harbormaster/internal/pkg/handler"
	"github.com/Nike-Inc/harbormaster/internal/pkg/validator"
	log "github.com/sirupsen/logrus"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"
)

var levelMapping = map[string]log.Level{
	"DEBUG": log.DebugLevel,
	"INFO":  log.InfoLevel,
	"WARN":  log.WarnLevel,
	"ERROR": log.ErrorLevel,
}

const (
	portFlag              = "port"
	healthPortFlag        = "health-port"
	logLevelFlag          = "log-level"
	caCertFlag            = "ca-cert"
	certFlag              = "cert"
	keyFlag               = "key"
	clientIDFlag          = "client-id"
	issuerURLFlag         = "issuer-url"
	usernameClaimFlag     = "username-claim"
	groupClaimFlag        = "group-claim"
	memoryFlag            = "cache-max-memory"
	cacheExpiryFlag       = "cache-expiry"
	graphClientIDFlag     = "graph-client-id"
	graphClientSecretFlag = "graph-client-secret"
	graphTenantNameFlag   = "graph-tenant-name"
)

var harbormasterCmd = &cobra.Command{
	Use:   "harbormaster",
	Short: "A webhook handler for validating JWT",
	Long: `A webhook handler for validating JWT tokens sent to Kubernetes. In addition to
validating the token, it can also add data to the user information returned
to Kubernetes. This additional user information is cached to decrease the amount
of trips out to active directory. The cache duration and size are configurable.`,
	RunE: startHarbormaster,
}

func init() {
	viper.SetEnvPrefix("harbormaster")
	viper.AutomaticEnv()
	// This normalizes "-" to an underscore in env names.
	viper.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
	flags := harbormasterCmd.Flags()
	flags.IntP(portFlag, "p", 9000, "port harbormaster should listen on. [HARBORMASTER_PORT]")
	flags.IntP(memoryFlag, "m", validator.DefaultMaxCacheMemory, "amount of memory (in MB) to cap the cache at [HARBORMASTER_CACHE_MAX_MEMORY]")
	flags.DurationP(cacheExpiryFlag, "e", validator.DefaultCacheDuration, "how long before a cache entry is considered invalid [HARBORMASTER_CACHE_EXPIRY]")
	flags.Int(healthPortFlag, 8000, "port to listen on for health checks. [HARBORMASTER_HEALTH_PORT]")
	flags.StringP(logLevelFlag, "v", "WARN", "log level to use. Must be one of: 'DEBUG', 'INFO', 'WARN', or 'ERROR'. [HARBORMASTER_LOG_LEVEL]")
	flags.StringP(caCertFlag, "r", "/usr/share/harbormaster/ca.pem", "ca certificate to use for root trust. This is required. [HARBORMASTER_CA_CERT]")
	flags.StringP(certFlag, "c", "/usr/share/harbormaster/harbormaster.pem", "certificate to use for https. This is required. [HARBORMASTER_CERT]")
	flags.StringP(keyFlag, "k", "/usr/share/harbormaster/harbormaster.key", "keyfile for the certificate. This is required. [HARBORMASTER_KEY]")
	flags.StringP(clientIDFlag, "i", "", "OIDC provider client ID. [HARBORMASTER_CLIENT_ID]")
	flags.StringP(issuerURLFlag, "u", validator.DefaultIssuerURL, "OIDC token issuer URL. [HARBORMASTER_ISSUER_URL]")
	flags.StringP(usernameClaimFlag, "n", validator.DefaultUsernameClaim, "claim to use as the username [HARBORMASTER_USERNAME_CLAIM]")
	flags.StringP(groupClaimFlag, "g", validator.DefaultGroupsClaim, "claim to use for user's groups [HARBORMASTER_GROUP_CLAIM]")
	flags.String(graphClientIDFlag, "", "MS Graph application client ID to use [HARBORMASTER_GRAPH_CLIENT_ID]")
	flags.String(graphClientSecretFlag, "", "MS Graph application client secret to use [HARBORMASTER_GRAPH_CLIENT_SECRET]")
	flags.String(graphTenantNameFlag, "", "MS Graph application tenant name to use (generally <company>.onmicrosoft.com) [HARBORMASTER_GRAPH_TENANT_NAME]")
	viper.BindPFlags(flags)
}

// Given a location of a PEM encoded certificate and keyfile path, attempt to load
// the certificate and return a TLS config suitable for use with a server
func generateTLSConfig(rootCA string) (*tls.Config, error) {
	content, err := ioutil.ReadFile(rootCA)
	if err != nil {
		return nil, fmt.Errorf("Error reading root CA file: %s", err)
	}
	pool := x509.NewCertPool()
	if ok := pool.AppendCertsFromPEM(content); !ok {
		return nil, fmt.Errorf("Unable to parse ca file as a PEM encoded certificate")
	}
	conf := &tls.Config{
		ClientAuth: tls.RequireAndVerifyClientCert,
		MinVersion: tls.VersionTLS12,
		ClientCAs:  pool,
	}
	conf.BuildNameToCertificate()
	return conf, nil
}

func startHarbormaster(cmd *cobra.Command, args []string) error {
	if l, ok := levelMapping[viper.GetString(logLevelFlag)]; ok {
		log.SetLevel(l)
		f := &log.TextFormatter{
			FullTimestamp: true,
		}
		log.SetFormatter(f)
	} else {
		return fmt.Errorf("Invalid log level. Must be one of: 'DEBUG', 'INFO', 'WARN', or 'ERROR'")
	}

	if viper.GetString(clientIDFlag) == "" {
		return fmt.Errorf("--%s should be specified", clientIDFlag)
	}
	// Initialize the validator with all necessary information
	log.Debug("Setting up validator")
	var opts = []validator.Option{
		validator.ClientID(viper.GetString(clientIDFlag)),
		validator.IssuerURL(viper.GetString(issuerURLFlag)),
		validator.UsernameClaim(viper.GetString(usernameClaimFlag)),
		validator.GroupsClaim(viper.GetString(groupClaimFlag)),
		validator.CacheSize(viper.GetInt(memoryFlag)),
		validator.CacheExpiry(viper.GetDuration(cacheExpiryFlag)),
	}
	if viper.GetString(graphClientIDFlag) == "" || viper.GetString(graphClientSecretFlag) == "" || viper.GetString(graphTenantNameFlag) == "" {
		log.Warn("Graph client ID, client secret, or tenent name not set. Skipping group getter")
	} else {
		opts = append(opts, validator.GraphGetter(viper.GetString(graphClientIDFlag), viper.GetString(graphClientSecretFlag), viper.GetString(graphTenantNameFlag)))
	}
	v, err := validator.New(opts...)
	if err != nil {
		return err
	}
	log.Debug("Setting up OIDC handler")
	th := handler.NewTokenHandler(v)

	// Routing for webhook
	mux := http.NewServeMux()
	mux.HandleFunc("/auth/validate", th.ValidateTokenReview)

	// Generate the TLS configuration and set up the server
	listenAddr := fmt.Sprintf(":%d", viper.GetInt(portFlag))
	log.WithField("ca-path", viper.GetString(caCertFlag)).Debug("Generating TLS config with ca certificate")
	tlsConfig, err := generateTLSConfig(viper.GetString(caCertFlag))
	if err != nil {
		return err
	}
	s := http.Server{
		Addr:      listenAddr,
		TLSConfig: tlsConfig,
		Handler:   mux,
	}

	// Start the webhook server
	log.WithFields(log.Fields{
		"cert-path": viper.GetString(certFlag),
		"key-path":  viper.GetString(keyFlag),
		"address":   listenAddr,
	}).Info("Starting server")
	// TODO(taylor): Error handling on server startup
	go s.ListenAndServeTLS(viper.GetString(certFlag), viper.GetString(keyFlag))
	// Health check endpoint on non https
	var healthListenAddr = fmt.Sprintf(":%d", viper.GetInt(healthPortFlag))

	// Configure the health routing and start the health check server
	healthMux := http.NewServeMux()
	healthMux.HandleFunc("/healthz", handler.HealthCheck)
	log.WithField("address", healthListenAddr).Info("Starting health check endpoint")
	http.ListenAndServe(healthListenAddr, healthMux)
	return nil
}

func main() {
	if err := harbormasterCmd.Execute(); err != nil {
		os.Exit(1)
	}
}
